### ISSUE TITLE
Different inference results on Windows and Linux

### ISSUE THREAD

**Issue by tschauner-s** [ultralytics/yolov5#6936]
### Search before asking

- [x] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Detection

### No error message, just wrong results.

I cannot use my models (.pt file) trained on Windows machines on Linux servers. It does not have all parameters. Results are completely wrong.

- Windows (8.1 and 10) - correct results:

Fusing layers...
Model Summary: 290 layers, **20885262 parameters**, 0 gradients
Adding AutoShape...

- Ubuntu - wrong results:

Fusing layers...
Model Summary: 290 layers, **20865057 parameters**, 0 gradients
Adding AutoShape...

The number of parameters are not matching. Is there any solution?

### Environment

YOLOv5 🚀 v6.0-216-g856d4e57 torch 1.10.2+cpu CPU
Ubuntu / Windows 8.1, Windows 10
Python 3.9

### Minimal Reproducible Example

```
import torch
from cv2 import cv2`

yolov5_path = "C:/yolov5/"
model_path = "PATH/best.pt" # Model trained on Windows

img = cv2.imread("C:/xyz.jpg")

model = torch.hub.load(yolov5_path, 'custom', path=model_path, source='local', device='cpu', autoshape=True)

def detectObjects(input):
    results = model(input, size=640)
    results = results.pandas().xyxy[0]
    return results.to_dict(orient = 'index')

results = detectObjects(img)

for elem in results:
    print(elem, results[elem])
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@tschauner-s 👋 hi, thanks for letting us know about this possible problem with YOLOv5 🚀. We've created a few short guidelines below to help users provide what we need in order to get started investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been solved in [master](https://github.com/ultralytics/yolov5/tree/master).
- ✅ **Unmodified** – Your problem must be reproducible using official YOLOv5 code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) with a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by tschauner-s** [comment]
Still having the mentioned issue. Anyone having an idea what is going wrong?

**Comment by glenn-jocher** [comment]
@tschauner-s YOLOv5 is environment agnostic, so you can train a model anywhere and deploy it anywhere and expect the same results.

Identical results are not guaranteed across releases nor versions of dependencies. If you want a stable environment you might use our Docker Image.

Note that for training even when using the same seed trainings may produce different results, especially when using CUDA backends. See https://pytorch.org/docs/stable/notes/randomness.html for details on factors affecting PyTorch training reproducibility and sources of randomness.

<img width="949" alt="Screen Shot 2022-01-08 at 11 20 59 AM" src="https://user-images.githubusercontent.com/26833433/148660366-9155dfd9-351c-4174-818f-2b1b907b3e4e.png">

Good luck 🍀 and let us know if you have any other questions!

**Comment by tschauner-s** [comment]
@glenn-jocher Thanks for pointing this out. I highly doubt your environment agnosticism, as we used the same set of libraries and versions in both cases. Never ran into similar problems with any other PyTorch-based library.

> and deploy it anywhere and expect the same results.

That is the exact opposite of what I initially reported here.

I will keep you updated in case we find the problem.

**Comment by glenn-jocher** [comment]
@tschauner-s I understand your frustration and I appreciate your feedback. Your experience is valuable for both the YOLO community and the Ultralytics team. While environment agnosticism is a design goal, I acknowledge the unexpected inconsistency you encountered and look forward to any further updates you may have. Thank you for your contribution 🙏 and please feel free to reach out with any additional questions or concerns.

**Comment by wuhanstudio** [comment]
The same issue here. Different results on Windows and Linux.

**Comment by glenn-jocher** [comment]
@wuhanstudio, thanks for your message 🌟. It's crucial to ensure that the model, PyTorch, and CUDA versions are identical across both platforms for consistent results. Differences can sometimes originate from minor version discrepancies in dependencies or even the underlying hardware. Please double-check your environments, and if the issue persists, it may be helpful to share the specific model and version details so the community can investigate further. Let's keep collaborating to solve this puzzle! 🧩
